### ISSUE TITLE
Ambiguities in detokenization / what kinds of normalization are performed on the input?

### ISSUE THREAD

**Issue by prashantserai** [google/sentencepiece#766]
Hi! I am trying to use a pretrained Sentenpiece model and running into some interesting normalization issues.

The code below shows how the single character showing continuation (three dots) is turned into three normal "." characters. I also see multiple spaces being collapsed into a single space, addition of a leading whitespace, normalization of various emojis (Eg. double exclamation mark), and other interesting cases I observed.

Basically, I am no longer able to do the simple detokenization as shown on the homepage of the repo: detokenized = ''.join(pieces).replace('▁', ' ')

Is there perhaps a list or a reference for all the normalizations/magic performed by the sentencepiece library? Or is this a property of the sentencepiece model itself (that is provided by PyTorch)?

`import sentencepiece as spm`
`google_sp = spm.SentencePieceProcessor()`
`google_sp.load("xlmr.sentencepiece.bpe.model") # pretrained model from https://download.pytorch.org/models/text/xlmr.sentencepiece.bpe.model `
`text = "…" # problematic token examined individually`
`pieces = google_sp.EncodeAsPieces(text);`
`pieces = [piece.replace("▁", " ") for piece in pieces] # white space replacement`
`detokenized = "".join(pieces)[1:] # join and exclude the very first whitespace character`
>>> print(text, "\t", detokenized)
… 	 ...
>>> print(len(text), "\t", len(detokenized))
1 	 3

**Comment by prashantserai** [comment]
Kind of found the answer here:
https://github.com/google/sentencepiece/blob/master/doc/normalization.md

Is there an API (preferably in Python) to read a model file to verify what kind of normalization is going on?

**Comment by taku910** [comment]
The content of ``--nomalization_rule_name`` is stored in the model file. We can get the value as follows.

````
>>> import sentencepiece.sentencepiece_model_pb2 as model
>>> m = model.ModelProto()
>>> m.ParseFromString(open('m.model', 'rb').read())
274863
>>> m.normalizer_spec.name
'nmt_nfkc'
````

you can use ``spm_normalize`` command to extract the normalization rule as TSV.

```
 spm_normalize --decompile --model=m.model --output=norm.tsv
```
